### PR TITLE
feat: [PagoPA-1488] renewal debt positions

### PR DESCRIPTION
**Pull request by alessio-acitelli**
#### List of Changes

- created batch for archiving debit positions

#### Motivation and Context

- relieve the database of debt positions that are now obsolete

#### How Has This Been Tested?

- manual and junit test

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: helm/values-dev.yaml [hunk 1]
```diff
@@ -93,6 +93,15 @@ microservice-chart:
     CRON_JOB_SCHEDULE_ENABLED: "true"
     CRON_JOB_SCHEDULE_EXPRESSION_TO_VALID: "0 */10 * * * *"
     CRON_JOB_SCHEDULE_EXPRESSION_TO_EXPIRED: "0 */20 * * * *"
+    CRON_JOB_SCHEDULE_HISTORICIZATION_ENABLED: "true"
```
Comment (pasqualespica) [inline helm/values-dev.yaml:96]: all var name related to new feature ex.

- `CRON_JOB_SCHEDULE_HISTORY`+ `ENABLE`
- `CRON_JOB_SCHEDULE_HISTORY`+ `TRIGGER`
- `CRON_JOB_SCHEDULE_HISTORY`+ `QUERY`
...

File: helm/values-dev.yaml [hunk 2]
```diff
@@ -93,6 +93,15 @@ microservice-chart:
...
+    CRON_JOB_SCHEDULE_EXPRESSION_HISTORICIZATION_DEBT_POS: "0 */30 * * * *"
+    CRON_JOB_SCHEDULE_EXTRACTION_HISTORY_QUERY: "SELECT pp FROM PaymentPosition pp WHERE pp.status IN ('PAID', 'REPORTED', 'INVALID', 'EXPIRED') AND pp.lastUpdatedDate < ?1"
+    CRON_JOB_SCHEDULE_EXTRACTION_HISTORY_QUERY_PAGINATED: "true"
+    CRON_JOB_SCHEDULE_EXTRACTION_HISTORY_QUERY_COUNT: "SELECT count(pp.id) FROM PaymentPosition pp WHERE pp.status IN ('PAID', 'REPORTED', 'INVALID', 'EXPIRED') AND pp.lastUpdatedDate < ?1"
+    CRON_JOB_SCHEDULE_EXTRACTION_HISTORY_QUERY_PAGE_SIZE: "5"
+    CRON_JOB_SCHEDULE_EXTRACTION_HISTORY_QUERY_INTERVAL: "7"
+    ARCHIVE_SA_PO_TABLE: "pagopadweugpsarchivesapaymentoptiontable"
```
Comment (pasqualespica) [inline helm/values-dev.yaml:103]: @FedericoRuzzier remove prefix `pagopadweugpsarchives` into table name https://github.com/pagopa/pagopa-infra/pull/1711

File: helm/values-dev.yaml [hunk 3]
```diff
@@ -109,6 +118,7 @@ microservice-chart:
     SPRING_DATASOURCE_PASSWORD: "db-apd-user-password"
     SPRING_DATASOURCE_USERNAME: "db-apd-user-name"
     NODE_SUBSCRIPTION_KEY: "gpd-d-node-subscription-key"
+    ARCHIVE_SA_CONNECTION_STRING: "gpd-archive-d-sa-connection-string"
```
Comment (pasqualespica) [inline helm/values-dev.yaml:121]: add `GPD` prefix

File: src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    private boolean paginationMode;
+    @Value("${cron.job.schedule.extraction.history.query.count:SELECT count(pp.id) FROM PaymentPosition pp WHERE pp.status IN ('PAID', 'REPORTED', 'INVALID', 'EXPIRED') AND pp.lastUpdatedDate < ?1}")
+    private String countExtractionQuery;
+    @Value("${cron.job.schedule.extraction.history.query.page.size:100000}")
+    private int pageSize;
+    
+    // azure storage params
+    @Value("${azure.archive.storage.connection}")
+    private String archiveStorageConnection;
+    @Value("${azure.archive.storage.table.po:pagopadweugpsarchivesapaymentoptiontable}")
```
Comment (pasqualespica) [inline src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java:70]: oki, shall not contains env `pagopadweugpsarchives`

File: src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    private String archiveStoragePOTable;
+    @Value("${azure.archive.storage.table.pp:pagopadweugpsarchivesapaymentpositiontable}")
```
Comment (pasqualespica) [inline src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java:72]: ditto

File: src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+    @PersistenceUnit
+	private EntityManagerFactory emf;
+    
+    public HistoricizationScheduler(PaymentPositionRepository paymentPositionRepository) {
+		super();
+		this.paymentPositionRepository = paymentPositionRepository;
+	}
+  
+    @Scheduled(cron = "${cron.job.schedule.expression.historicization.debt.positions}")
+    @SchedulerLock(name = "HistoricizationScheduler_manageDebtPositionsToHistoricize", lockAtMostFor = "180m", lockAtLeastFor = "15m")
```
Comment (pasqualespica) [inline src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java:86]: should be usefull to customize it

File: src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    	log.info(String.format(LOG_BASE_HEADER_INFO, CRON_JOB, METHOD, "Running at " + DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss").format(LocalDateTime.now())));
+    	EntityManager em = this.getEntityManager();
+    	LocalDateTime ldt = LocalDateTime.now().minusDays(extractionInterval);
+    	List<PaymentPosition> ppList;
+    	if (paginationMode) {
+    		long countResult = em.createQuery(countExtractionQuery, Long.class).setParameter(1,ldt).getSingleResult(); 
+    		int numOfPages = (int)Math.ceil((float)countResult/pageSize);
+    		for (int pageNumber=0; pageNumber<numOfPages; pageNumber++) {
+    			ppList = em.createQuery(extractionQuery, PaymentPosition.class).setParameter(1,ldt).setFirstResult((pageNumber) * pageSize).setMaxResults(pageSize).getResultList();
+    			log.info(String.format(LOG_BASE_HEADER_INFO, CRON_JOB, METHOD,
```
Comment (pasqualespica) [inline src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java:99]: after test reduce it to `debug`

File: src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+				tableClient.updateEntity(tableEntity);
+			} else {
+				log.error(String.format(LOG_BASE_HEADER_INFO, CRON_JOB, "saveToPOTable",
+						"error while storing the table information [organizationFiscalCode="+organizationFiscalCode+", iuv="+po.getIuv()+"]"), e);
+				throw e;
+			}
+		}
+	}
+
+	public void saveToPPTable(String organizationFiscalCode, PaymentPosition pp, ObjectMapper objectMapper) throws JsonProcessingException {
```
Comment (pasqualespica) [inline src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java:151]: change name to `upsert`

File: src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    }
+    
+    public TableClient getTableClient(String connectionString, String tableName) {
+    	return new TableClientBuilder()
+			    .connectionString(connectionString)
+			    .tableName(tableName)
+			    .buildClient();
+    }
+
+	public void saveToPOTable(String organizationFiscalCode, PaymentPosition pp, PaymentOption po) {
```
Comment (pasqualespica) [inline src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java:129]: change name to `upsert`

File: src/main/resources/db/migration/V011__ADD_SHEDLOCK.sql [hunk 10]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE shedlock (
```
Comment (pasqualespica) [inline src/main/resources/db/migration/V011__ADD_SHEDLOCK.sql:1]: add comment/ref doc

**Review by pasqualespica — CHANGES_REQUESTED** [review]
great 🚀

File: helm/values-dev.yaml (hunk 2 above)
Comment (alessio-acitelli) [reply helm/values-dev.yaml:103]: done

File: helm/values-dev.yaml (hunk 3 above)
Comment (alessio-acitelli) [reply helm/values-dev.yaml:121]: done

File: src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java (hunk 4 above)
Comment (alessio-acitelli) [reply src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java:70]: done

File: src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java (hunk 5 above)
Comment (alessio-acitelli) [reply src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java:72]: done

File: src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java (hunk 7 above)
Comment (alessio-acitelli) [reply src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java:99]: TODO after test in UAT

File: src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java (hunk 8 above)
Comment (alessio-acitelli) [reply src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java:151]: done

File: src/main/resources/db/migration/V011__ADD_SHEDLOCK.sql (hunk 10 above)
Comment (alessio-acitelli) [reply src/main/resources/db/migration/V011__ADD_SHEDLOCK.sql:1]: done

File: src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java (hunk 6 above)
Comment (alessio-acitelli) [reply src/main/java/it/gov/pagopa/debtposition/scheduler/HistoricizationScheduler.java:86]: done

**Review by pasqualespica — APPROVED** [review]
manual test too https://pagopa.atlassian.net/wiki/spaces/IQCGJ/pages/926451650/Test+svecchiamento+in+DEV+12-feb-2024

tnx2 @AngeloCaporaso @alessio-acitelli

File: helm/values-dev.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (alessio-acitelli) [reply helm/values-dev.yaml]: done

**Pull request merged**